src/get-jwks.js

```diff
@@ -4,31 +4,51 @@ const fetch = require('node-fetch')
 const lru = require('tiny-lru')
 const jwkToPem = require('jwk-to-pem')
 
-const MISSING_JWK_ERROR = 'No matching JWK found in the set.'
-const NO_JWKS_ERROR = 'No JWKS found in the response.'
+const errors = {
+  NO_JWKS: 'No JWKS found in the response.',
+  JWK_NOT_FOUND: 'No matching JWK found in the set.',
+  DOMAIN_NOT_ALLOWED: 'The domain is not allowed.'
+}
+
+function ensureTrailingSlash (domain) {
+  return domain.endsWith('/') ? domain : `${domain}/`
+}
+
+function buildGetJwks (options = {}) {
+  const max = options.max || 100
+  const ttl = options.ttl || 60 * 1000
+  const allowedDomains = (options.allowedDomains || []).map(ensureTrailingSlash)
 
-function buildGetJwks (cacheProps = {}) {
-  const max = cacheProps.max || 100
-  const ttl = cacheProps.ttl || 60 * 1000
   const cache = lru(max, ttl)
 
-  async function getPublicKey (signatures) {
-    const key = await getJwk(signatures)
-    const publicKey = jwkToPem(key)
-    return publicKey
+  async function getPublicKey (signature) {
+    return jwkToPem(await this.getJwk(signature))
```

Comment:
You would need to add a `catch` block that access `err.stack`  to avoid a memory leak.
Read up on https://bugs.chromium.org/p/v8/issues/detail?id=4714 and https://bugs.chromium.org/p/v8/issues/detail?id=2340.

A better approach is to not store a rejected promise in the cache.

Reply:
if we do not store the rejected promise are we not risking a DOS if we get tokens that always cause retrieval errors because we're hitting the network each single time? I'm not familiar with the memory leak you mention so I'll have to read up on it

Reply:
You could also access `err.stack` and remove the problem.

You are right about the network problem, that's why I suggested #20.

Reply:
see all the updates, I batched the 3 issues in this PR. as for rejected promises:

- we store the unsettled promise in the cache to avoid duplicate requests
- once the promise settles, if it rejected we remove it from the cache

I believe this whole thing can really get arbitrarily more complex and I'm not sure it's really worth it, let's consider that we're simply doing a request to an API which returns a bunch of keys and that's all. I think we could get away with a much simpler mechanism to be honest